File: app/src/main/java/com/manoj/dlt/features/DeepLinkHistoryFeature.java
package com.manoj.dlt.features;

import android.content.Context;

import com.google.firebase.database.DatabaseReference;
import com.manoj.dlt.Constants;
import com.manoj.dlt.DbConstants;
import com.manoj.dlt.events.DeepLinkFireEvent;
import com.manoj.dlt.interfaces.IDeepLinkHistory;
import com.manoj.dlt.models.DeepLinkInfo;
import com.manoj.dlt.models.ResultType;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DeepLinkHistoryFeature implements IDeepLinkHistory
{
    private static DeepLinkHistoryFeature _instance;
    private FileSystem _fileSystem;
    private Context _context;

    private DeepLinkHistoryFeature(Context context)
    {
        _fileSystem = new FileSystem(context, Constants.DEEP_LINK_HISTORY_KEY);
        _context = context;
        migrateHistoryToFirebase();
        EventBus.getDefault().register(this);
    }

    public static DeepLinkHistoryFeature getInstance(Context context)
    {
        if(_instance == null)
        {
            _instance = new DeepLinkHistoryFeature(context);
        }
        return _instance;
    }

    @Override
    public void addLinkToHistory(final DeepLinkInfo deepLinkInfo)
    {
        DatabaseReference baseUserReference = ProfileFeature.getInstance(_context).getCurrentUserFirebaseBaseRef();
        DatabaseReference linkReference = baseUserReference.child(DbConstants.USER_HISTORY).child(deepLinkInfo.getId());
        Map<String, Object> infoMap = new HashMap<String, Object>(){{
        put(DbConstants.DL_ACTIVITY_LABEL, deepLinkInfo.getActivityLabel());
        put(DbConstants.DL_DEEP_LINK, deepLinkInfo.getDeepLink());
        put(DbConstants.DL_PACKAGE_NAME, deepLinkInfo.getPackageName());
        put(DbConstants.DL_UPDATED_TIME, deepLinkInfo.getUpdatedTime());
        }};
        linkReference.setValue(infoMap);
    }

    @Override
    public void removeLinkFromHistory(String deepLinkId)
    {
        DatabaseReference baseUserReference = ProfileFeature.getInstance(_context).getCurrentUserFirebaseBaseRef();
        DatabaseReference linkReference = baseUserReference.child(DbConstants.USER_HISTORY).child(deepLinkId);
        linkReference.setValue(null);
    }

    @Override
    public void clearAllHistory()
    {
        DatabaseReference baseUserReference = ProfileFeature.getInstance(_context).getCurrentUserFirebaseBaseRef();
        DatabaseReference historyRef = baseUserReference.child(DbConstants.USER_HISTORY);
        historyRef.setValue(null);
    }

    @Subscribe(sticky = true, priority = 1)
    public void onEvent(DeepLinkFireEvent deepLinkFireEvent)
    {
        if(deepLinkFireEvent.getResultType().equals(ResultType.SUCCESS))
        {
            addLinkToHistory(deepLinkFireEvent.getDeepLinkInfo());
        }
    }

    //legacy function to migrate db to firebase, for older version
    //TODO: remove once all users migrated
    private void migrateHistoryToFirebase()
    {
        for(DeepLinkInfo info: getLinkHistoryFromFileSystem())
        {
            addLinkToHistory(info);
        }
        _fileSystem.clearAll();
    }

    private List<DeepLinkInfo> getLinkHistoryFromFileSystem()
    {
        List<DeepLinkInfo> deepLinks = new ArrayList<DeepLinkInfo>();
        for (String deepLinkInfoJson : _fileSystem.values())
        {
            deepLinks.add(DeepLinkInfo.fromJson(deepLinkInfoJson));
        }
        Collections.sort(deepLinks);
        return deepLinks;
    }
}

File: app/src/main/kotlin/com/thunderclouddev/deeplink/features/DeepLinkHistoryFeature.kt
package com.thunderclouddev.deeplink.features

import android.content.Context
import com.thunderclouddev.deeplink.BaseApplication
import com.thunderclouddev.deeplink.Constants
import com.thunderclouddev.deeplink.DbConstants
import com.thunderclouddev.deeplink.events.DeepLinkFireEvent
import com.thunderclouddev.deeplink.interfaces.IDeepLinkHistory
import com.thunderclouddev.deeplink.models.DeepLinkInfo
import com.thunderclouddev.deeplink.models.ResultType
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import java.util.*

class DeepLinkHistoryFeature private constructor(private val _context: Context) : IDeepLinkHistory {
    private val _fileSystem: FileSystem

    init {
        _fileSystem = FileSystem(_context, Constants.DEEP_LINK_HISTORY_KEY)
        EventBus.getDefault().register(this)
    }

    override fun addLinkToHistory(deepLinkInfo: DeepLinkInfo) {
        BaseApplication.database.putLink(deepLinkInfo)
    }

    override fun removeLinkFromHistory(deepLinkId: String) {
        BaseApplication.database.removeLink(deepLinkId)
    }

    override fun clearAllHistory() {
        BaseApplication.database.clearAllHistory()
    }

    override val linkHistoryFromFileSystem: List<DeepLinkInfo>
        get() {
            return _fileSystem.values()
                    .mapNotNull { DeepLinkInfo.fromJson(it) }
                    .sorted()
        }

    @Subscribe(sticky = true, priority = 1)
    fun onEvent(deepLinkFireEvent: DeepLinkFireEvent) {
        if (deepLinkFireEvent.resultType == ResultType.SUCCESS) {
            addLinkToHistory(deepLinkFireEvent.info)
        }
    }

    private fun clearFileSystemHistory() {
        _fileSystem.clearAll()
    }

    companion object {
        private var _instance: DeepLinkHistoryFeature? = null

        fun getInstance(context: Context): DeepLinkHistoryFeature {
            if (_instance == null) {
                _instance = DeepLinkHistoryFeature(context)
            }

            return _instance as DeepLinkHistoryFeature
        }
    }
}
